File: internal/calendar/month.go
package calendar

import (
	"strings"
	"time"
)

// MonthFormat is a month as the front page's rail asks for one: "2026-09".
const MonthFormat = "2006-01"

// Month is a month as Heliosian's rail shows it to one viewer: today, the
// school days in the month with what kind of day each is for their
// classrooms, and every event of theirs that touches the month - what the
// small month's dots and the day card under it are drawn from.
type Month struct {
	Month string `json:"month"`
	Today string `json:"today"`
	// Days holds every school day in the month, by date, with the day types
	// in force for the viewer's classrooms other than Regular; a date that
	// is not here is a weekend, a holiday the year calendar leaves out, or
	// outside the school year.
	Days   map[string]Day `json:"days"`
	Events []Upcoming     `json:"events"`
}

// Day is one school day: the day types in force for the viewer's
// classrooms other than Regular, each naming its classrooms when they are
// not all of them - "Early Dismissal", "No Aftercare · Hummingbirds" - the
// way the calendar's own day words read.
type Day struct {
	Kinds []Kind `json:"kinds"`
}

// Kind is one day type in force: its name, and the words for it.
type Kind struct {
	Name  string `json:"name"`
	Words string `json:"words"`
}

// Month reckons a month for one viewer under the filter the calendar page
// first shows them (viewOf): the plan for their classrooms day by day, and
// the events the view admits that fall in the month, in date order, each
// with the viewer's answer and the ones they hid left out. A month that
// does not parse is the month now is in.
func (m *Model) Month(directory Directory, email string, linked []Linked, now time.Time, month string) Month {
	return m.MonthUnder(directory, email, linked, now, month, "")
}

// MonthUnder is Month read under one of the person's saved calendars by
// token, or My Heliosian by its token, rather than their default - the
// rail's picker - or under the default for a blank or unknown token.
func (m *Model) MonthUnder(directory Directory, email string, linked []Linked, now time.Time, month, token string) Month {
	first, err := time.ParseInLocation(MonthFormat, month, Location)
	if err != nil {
		first = time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, Location)
	}
	last := first.AddDate(0, 1, -1)
	from, to := first.Format(DateFormat), last.Format(DateFormat)
	classrooms, tags := m.viewUnder(directory, email, token)
	out := Month{Month: first.Format(MonthFormat), Today: now.Format(DateFormat), Days: map[string]Day{}, Events: []Upcoming{}}
	for d := first; !d.After(last); d = d.AddDate(0, 0, 1) {
		date := d.Format(DateFormat)
		byClassroom, ok := m.Days[date]
		if !ok {
			continue
		}
		byType := map[string][]string{}
		for _, c := range classrooms {
			if name := byClassroom[c]; name != "" && name != RegularDayType {
				byType[name] = append(byType[name], c)
			}
		}
		day := Day{Kinds: []Kind{}}
		for _, t := range m.DayTypes {
			rooms, ok := byType[t.Name]
			if !ok {
				continue
			}
			words := t.Name
			if len(rooms) != len(classrooms) {
				words += " · " + strings.Join(rooms, ", ")
			}
			day.Kinds = append(day.Kinds, Kind{Name: t.Name, Words: words})
		}
		out.Days[date] = day
	}
	for _, e := range m.eventsFor(email, linked) {
		answer := m.AnswerOf(email, e.ID)
		if e.start.Format(DateFormat) > to || e.end.Format(DateFormat) < from || answer == AnswerHidden || !admits(m, e, classrooms, tags) {
			continue
		}
		u := m.card(e)
		u.Answer = answer
		out.Events = append(out.Events, u)
	}
	return out
}

File: web/home/month.js
import {state} from './state.js';
import {el, svg} from './dom.js';
import {whenOrigin, rsvpButtons} from './cards.js';

// The rail's calendar, from Helios When: a small month, paged on its own,
// with a dot under each day in the colour of what is on it and today ringed
// in amber, and under it the day picked - today until one is - as a card:
// the date, what kind of day it is for this person's classrooms when it is
// not simply regular, and every event on it, each opening its page on When
// with its Yes and No under it.
// The month is the viewer's as When first shows it (their classrooms, the
// default categories); today is the school's, reckoned by the server, so
// the ring does not drift with the browser's clock.

let month = null;
let selected = '';

// A YYYY-MM-DD as a local date, without the time zone shifting it.
function parseDate(date) {
  const [y, m, d] = date.split('-').map(Number);
  return new Date(y, m - 1, d);
}

const pad = n => String(n).padStart(2, '0');

function dateOf(d) {
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}`;
}

const dayTypeClass = {'No School': 'dt-no-school', 'Early Dismissal': 'dt-early', 'No Aftercare': 'dt-no-aftercare'};

// tint is where an event comes from, as When colours it: a party pink, an
// HCA event purple, the school's own blue.
function tint(event) {
  return event.linkApp === 'celebrate' ? 'is-celebrate' : event.linkApp === 'team' ? 'is-team' : 'is-school';
}

function lastDay(event) {
  return (event.endAt || event.startAt).slice(0, 10);
}

function eventsOn(date) {
  return (month.events || []).filter(e => e.start <= date && lastDay(e) >= date);
}

// hours are what a row says beside the title: the hours the event's when
// line carries, else how far a whole-day event runs - "All day", or
// "Through Fri" while it has days to go.
function hours(event, date) {
  const [, time] = event.when.split(' · ');
  if (time) {
    return time;
  }
  const last = lastDay(event);
  if (last > date) {
    return 'Through ' + parseDate(last).toLocaleDateString('en-US', {weekday: 'short'});
  }
  return 'All day';
}

function monthLabel(ym) {
  return parseDate(ym + '-01').toLocaleDateString('en-US', {month: 'long', year: 'numeric'});
}

function shiftMonth(ym, by) {
  const d = parseDate(ym + '-01');
  d.setMonth(d.getMonth() + by);
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}`;
}

// Paging asks the server for the month, read for the viewer the way the
// first was; the day picked becomes today when it is in the month, else
// the first.
async function page(by) {
  const ym = shiftMonth(month.month, by);
  try {
    const res = await fetch('/api/apps/calendar?month=' + ym);
    if (!res.ok) {
      return;
    }
    month = await res.json();
  } catch {
    return;
  }
  selected = month.today.startsWith(month.month) ? month.today : month.month + '-01';
  renderMonth();
}

function grid() {
  const wrap = el('div', 'mini');
  const head = el('div', 'mini-head');
  const back = el('button', 'mini-page');
  back.type = 'button';
  back.setAttribute('aria-label', 'Previous month');
  back.append(svg('chevron'));
  back.addEventListener('click', () => page(-1));
  const next = el('button', 'mini-page');
  next.type = 'button';
  next.setAttribute('aria-label', 'Next month');
  next.append(svg('chevron'));
  next.addEventListener('click', () => page(1));
  head.append(back, el('span', 'mini-title', monthLabel(month.month)), next);
  wrap.append(head);
  const days = el('div', 'mini-grid');
  for (const w of ['S', 'M', 'T', 'W', 'T', 'F', 'S']) {
    days.append(el('span', 'mini-weekday', w));
  }
  const first = parseDate(month.month + '-01');
  for (let i = 0; i < first.getDay(); i++) {
    days.append(el('span', 'mini-blank'));
  }
  const last = new Date(first.getFullYear(), first.getMonth() + 1, 0).getDate();
  for (let n = 1; n <= last; n++) {
    const date = `${month.month}-${pad(n)}`;
    const day = month.days[date];
    const cell = el('button', 'mini-day');
    cell.type = 'button';
    if (date === month.today) {
      cell.classList.add('is-today');
    }
    if (date === selected) {
      cell.classList.add('is-selected');
    }
    if (!day) {
      cell.classList.add('is-off');
    } else if (day.kinds.length) {
      cell.classList.add(dayTypeClass[day.kinds[0].name] || 'dt-other');
    }
    cell.append(el('span', 'mini-number', String(n)));
    const dots = el('span', 'mini-dots');
    const seen = new Set();
    for (const e of eventsOn(date)) {
      const t = tint(e);
      if (!seen.has(t)) {
        seen.add(t);
        dots.append(el('span', 'mini-dot ' + t));
      }
    }
    cell.append(dots);
    cell.addEventListener('click', () => {
      selected = date;
      renderMonth();
    });
    days.append(cell);
  }
  wrap.append(days);
  return wrap;
}

function dayCard() {
  const card = el('div', 'rail-day');
  const date = parseDate(selected);
  const head = el('a', 'rail-day-head');
  head.href = whenOrigin('calendar') + '/day/' + selected;
  head.append(el('span', 'rail-day-title', selected === month.today ? 'Today' : date.toLocaleDateString('en-US', {weekday: 'long'})));
  head.append(el('span', 'rail-day-date', date.toLocaleDateString('en-US', {month: 'long', day: 'numeric'})));
  card.append(head);
  const day = month.days[selected];
  if (day && day.kinds.length) {
    const kinds = el('div', 'rail-day-kinds');
    for (const kind of day.kinds) {
      kinds.append(el('span', 'rail-day-kind ' + (dayTypeClass[kind.name] || 'dt-other'), kind.words));
    }
    card.append(kinds);
  }
  const events = eventsOn(selected);
  const list = el('div', 'rail-day-events');
  if (!events.length) {
    list.append(el('div', 'rail-day-empty', day ? 'Nothing on the calendar.' : 'No school.'));
  }
  // Each event is its row, opening its page on When, with its answer's
  // buttons under it - the same Yes and No the Upcoming cards carry.
  for (const event of events) {
    const item = el('div', 'rail-item');
    const row = el('a', 'rail-event ' + tint(event));
    row.href = whenOrigin('calendar') + event.path;
    row.append(el('span', 'rail-event-dot'));
    const body = el('span', 'rail-event-body');
    body.append(el('span', 'rail-event-title', event.title));
    body.append(el('span', 'rail-event-hours', hours(event, selected)));
    row.append(body);
    const rsvp = el('div', 'rail-rsvp');
    rsvp.append(rsvpButtons(event));
    item.append(row, rsvp);
    list.append(item);
  }
  card.append(list);
  return card;
}

export function renderMonth() {
  const root = document.querySelector('#rail-calendar');
  root.replaceChildren();
  if (!month) {
    month = state.model.calendar;
    if (!month || !month.month) {
      return;
    }
    selected = month.today;
  }
  root.append(grid(), dayCard());
}
